### ISSUE TITLE
Improve frame synchronization

### ISSUE THREAD

**Issue by theyareonit** [jellysquid3/Sodium#2610]
### Request Description

The function RenderSystem.limitDisplayFPS() in vanilla Minecraft currently looks something like this:
```
    public static void limitDisplayFPS(int fps) {
        double target = lastDrawTime + 1.0 / (double)fps;

        double now;
        for (now = GLFW.glfwGetTime(); now < target; now = GLFW.glfwGetTime()) {
            GLFW.glfwWaitEventsTimeout(target - now);
        }

        lastDrawTime = now;
    }
```
However, this is a flawed implementation. For example, if you set 240FPS as your cap, it will result in an actual frame limit of around ~235FPS, due to the fact that glfwWaitEventsTimeout() and glfwGetTime() do not return instantly, resulting in the `now` timestamp differing from the `target` timestamp by enough to cause significant desync.

The solution is simply to replace the line `lastDrawTime = now;` with `lastDrawTime = target;`, which results in a smoother experience when using capped FPS. However, this leads to an incorrect frame limit for the first few seconds after joining a server or tabbing into the game, so some more work may be needed.

**Comment by theyareonit** [comment]
There are also some extremely minor gains to be had from using Instant.now() as a timer instead if we're in a microoptimizing mood:
```
    public static void limitDisplayFPS(int fps) {
        target = target.plusNanos(1000000000 / fps);

        Instant now = Instant.now();
        for (; now.isBefore(target); now = Instant.now()) {}
        GLFW.glfwPollEvents();
    }

    private static Instant target = Instant.now();
```

![export](https://github.com/user-attachments/assets/a0bbe448-924b-4d33-8a2c-186c71176d53)

Though I'm not sure how well this will carry over to other platforms, I only tested this on Linux.

**Comment by jellysquid3** [comment]
That's curious. Since you appear to have figured this out already, we would not be opposed to a pull request to address this problem. Maybe we can also look into further improvements since the original Minecraft code seems rather naive.

**Comment by theyareonit** [comment]
Alright. I found some places where I can improve it (and fixed the issue of getting the wrong FPS value after tabbing in), so I'll try to make a PR soon.

**Comment by Felix14-v2** [comment]
If I understand correctly, this bug is quite useful for me. My laptop supports a screen refresh rate of up to 144 Hz, but in Minecraft I can only use a 150 FPS limit to make the game smooth. Due to this effect, I get ≈144 FPS on average.

**Comment by theyareonit** [comment]
> If I understand correctly, this bug is quite useful for me. My laptop supports a screen refresh rate of up to 144 Hz, but in Minecraft I can only use a 150 FPS limit to make the game smooth. Due to this effect, I get ≈144 FPS on average.

I think that just means some mod needs to allow the user to input any FPS value instead, lol.

**Comment by jellysquid3** [comment]
> If I understand correctly, this bug is quite useful for me. 

<img src="https://github.com/user-attachments/assets/b983cbb9-169f-409b-ac1e-95e780db3819" height="350" />

**Comment by Felix14-v2** [comment]
Yeah, so true. Just hope that one day modern screen formats support in Minecraft will be better.

**Comment by xboct02** [comment]
light vsync mod
